# MBA em Ciência de Dados
## Técnicas Avançadas de Captura e Tratamento de Dados


### <span style="color:darkred">Módulo III - Aquisição e Transformação de Dados</span>


## <span style="color:darkred">Avaliação</span>

Moacir Antonelli Ponti

CeMEAI - ICMC/USP São Carlos

---
#### <font color='red'> As respostas devem ser fornecidas no Moodle. O notebook é apenas para a implementação dos códigos que fornecerão as respostas</font>

---

Utilize as bibliotecas e carrege os dados conforme descrito abaixo

In [1]:
# carregando as bibliotecas necessárias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

escolas = pd.read_csv('ALUNOS_ANALISE.csv', header=None, sep=';')

<IPython.core.display.Javascript object>

In [2]:
escolas = escolas.iloc[:, 0:21]
escolas.columns = ['CDREDE', 'DE', 'CODMUN', 'MUN', 'CATEG', 'COD_ESC', 'TIPOESC', 'CODVINC', 'NOMESC', 'ENDESC', 'NUMESC', 'BAIESC', 'EMAIL', 'FONE1', 'ZONA', 'ED_INFANTIL', 'CLASSES ESPECIAIS', 'SALA DE RECURSO', 'ANOS INICIAIS', 'ANOS FINAIS', 'ENSINO MEDIO']

escolas.head()

CDREDE       DE  CODMUN        MUN  CATEG  COD_ESC  TIPOESC  CODVINC  \
0   10101  NORTE 1     100  SAO PAULO      1       12        8       12   
1   10101  NORTE 1     100  SAO PAULO      1       24        8       24   
2   10101  NORTE 1     100  SAO PAULO      1       36        8       36   
3   10101  NORTE 1     100  SAO PAULO      1       48        8       48   
4   10101  NORTE 1     100  SAO PAULO      1       59        8       59   

                                   NOMESC                     ENDESC  ...  \
0                AYRES DE MOURA PROFESSOR          RUA ARTUR ORLANDO  ...   
1               GAVIAO PEIXOTO BRIGADEIRO                   MOGEIRO   ...   
2                            JOAO SOLIMEO                   DO SABAO  ...   
3       WALFREDO ARANTES CALDAS PROFESSOR  DEPUTADO CANTIDIO SAMPAIO  ...   
4  SEBASTIAO DE OLIVEIRA GUSMAO PROFESSOR        RUA FRANCISCO GIRON  ...   

             BAIESC                        EMAIL       FONE1  ZONA  \
0      VILA JAGUARA  E000012A@EDUCACAO.SP.GOV.BR  36251633.0     1   
1        VILA PERUS  E000024A@EDUCACAO.SP.GOV.BR  39156810.0     1   
2  JARDIM MARISTELA  E000036A@EDUCACAO.SP.GOV.BR  39228491.0     1   
3        VILA SOUZA  E000048A@EDUCACAO.SP.GOV.BR  38513005.0     1   
4        MORRO DOCE  E000059A@EDUCACAO.SP.GOV.BR  39111913.0     1   

   ED_INFANTIL  CLASSES ESPECIAIS  SALA DE RECURSO  ANOS INICIAIS  \
0            0                  0                0              0   
1            0                  0                0            523   
2            0                  0               26              0   
3            0                  0                0              0   
4            0                  0                0              0   

   ANOS FINAIS  ENSINO MEDIO  
0          562             0  
1          473          1073  
2         1204           932  
3            0          1140  
4            0          1071  

[5 rows x 21 columns]

### Questão 1)
 
Porque é importante auditar fontes de dados já existentes e utilizar método e planejamento para realizar coleta de dados?

(a) para evitar analisar dados com viés e chegar a conclusões inválidas<br>
(b) para obter acurácias maiores<br>
(c) para evitar analisar dados errôneos e assim obter estimadores com menos erro<br>
(d) para evitar usar evidências anedotais para chegar a conclusões<br>


#### RESPOSTA:<br>**LETRA A**

### Questão 2)
 
Acesse o portal : http://catalogo.governoaberto.sp.gov.br/

Procure pelo arquivo CSV relativo a "Quantidade de alunos por tipo de ensino da rede estadual - 01/2019" (Secretaria da Educação - Sede) https://dados.educacao.sp.gov.br/dataset/quantidade-de-alunos-por-tipo-de-ensino-da-rede-estadual

Carregue os dados (considere as particularidades do arquivo e não carregue o cabeçalho). Depois, remova as colunas 21 em diante, mantendo as colunas de 0 a 20. Essas colunas restantes possuem significado de acordo com o "dicionário de dados" disponível ao visualizar o recurso dos dados. Sendo rotuladas da seguinte forma:
- CDREDE
- DE
- CODMUN
- MUN
- CATEG
- COD_ESC
- TIPOESC
- CODVINC
- NOMESC
- ENDESC
- NUMESC
- BAIESC
- EMAIL
- FONE 1
- ZONA
- ED_INFANTIL
- CLASSES ESPECIAIS
- SALA DE RECURSO
- ANOS INICIAIS
- ANOS FINAIS
- ENSINO MEDIO

Após carregar os dados e nomear as colunas, crie um novo atributo 'TOT' por meio da soma das 6 últimas colunas: ED_INFANTIL, CLASSES ESPECIAIS, SALA DE RECURSO, ANOS INICIAIS, ANOS FINAIS e ENSINO MEDIO.

Atribua nulo (nan) aos elementos cujo total (TOT) é zero.

In [3]:
escolas['TOT'] = escolas.iloc[:, 15:21].sum(axis=1)
escolas['TOT'] = escolas['TOT'].replace(0, np.nan)

<IPython.core.display.Javascript object>

Realize a **discretização** da variável 'TOT' utilizando:

1. o método dos quantis, utilizando 5 valores alvo, relativos aos quantis 0, 20, 40, 60, 80, 100

2. o método dos intervalos, utilizando 5 intervalos alvo: (0, 10] (10, 50] (50, 100] (100, 500] (500, max(TOT)],
em que max(TOT) é o maior valor desse atributo

Use o método qcut() para o item 1 e cut() para o item 2

Adicione essa novas variáveis na base, com os nomes 'TOT_5Q' (quantis) e 'TOT_5I' (intervalos)

In [5]:
bins = [0, 10, 50, 100, 500, escolas['TOT'].max()]

escolas['TOT_5Q'] = pd.qcut(escolas['TOT'], q=[0, .2, .4, .6, .8, 1])
escolas['TOT_5I'] = pd.cut(escolas['TOT'], bins=bins)


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [6]:
escolas['TOT_5Q'].value_counts()

(0.999, 323.0]     1051
(670.0, 918.0]     1045
(494.0, 670.0]     1045
(918.0, 2269.0]    1041
(323.0, 494.0]     1038
Name: TOT_5Q, dtype: int64

In [7]:
escolas['TOT_5I'].value_counts()

(500.0, 2269.0]    3099
(100.0, 500.0]     1898
(50.0, 100.0]       131
(10.0, 50.0]         74
(0.0, 10.0]          18
Name: TOT_5I, dtype: int64

Qual é a quantidade de dados (frequências dos valores discretizados) na base após a discretização, relativos ao primeiro intervalo e ao último intervalo para, respectivamente, TOT_5Q e TOT_5I?

(a) TOT_5Q: primeiro 0.999, último 2269. TOT_5I: primeiro 0, último 2269<br>
(b) TOT_5Q: primeiro 1051, último 1038. TOT_5I: primeiro 3099, último 18<br>
(c) TOT_5Q: primeiro 0.999, último 323. TOT_5I: primeiro 500, último 2269<br>
(d) TOT_5Q: primeiro 1051, último 1041. TOT_5I: primeiro 18, último 3099.<br>


#### RESPOSTA:<br> **LETRA B**

### Questão 3)

Normalize 2 variáveis da base: TOT e ENSINO MEDIO
* ENSINO MEDIO utilizando normalização da média
* TOT utilizando normalização L-2

Para isso, codifique funções que recebam uma coluna por parâmetro e retornem um atributo já normalizado

Depois, aplique as funções e crie novas variáveis com os atributos normalizados: MEDIO_nm e TOT_l2.

Após normalização, quais os valores de  média, desvio padrão, mínimo e máximo dessas variáveis, arredondando para 3 casas decimais?

(a) MEDIO_nm: 0.000, 0.132, 0.000, 0.876; TOT_l2: 0.012, 0.007, 0.000, 1.000<br>
(b) MEDIO_nm: 0.000, 0.132, -0.124, 0.876; TOT_l2: 0.012, 0.007, 0.000, 0.043<br>
(c) MEDIO_nm: 1.000, 0.132, 0.000, 2.000; TOT_l2: 0.012, 0.007, 0.000, 1.000<br>
(d) MEDIO_nm: 0.000, 0.500, -1.124, 1.125; TOT_l2: 0.012, 0.007, 0.000, 0.043<br>



In [8]:
def scaler_mean(x):
    return (x - x.mean()) / (x.max() - x.min())

def scaler_norm(x, norm='l2'):
    if norm == 'l1':  return x / np.abs(x).sum()
    if norm == 'l2':  return x / np.sqrt(np.sum(x ** 2))
    if norm == 'max': return x / x.max()
    return x

def describe(x):
    print('mean {:.3f}'.format(x.mean()))
    print('std  {:.3f}'.format(x.std()))
    print('min  {:.3f}'.format(x.min()))
    print('max  {:.3f}'.format(x.max()))

In [9]:
escolas['MEDIO_nm'] = scaler_mean(escolas['ENSINO MEDIO'])

In [10]:
escolas['TOT_l2'] = scaler_norm(escolas['TOT'])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [11]:
describe(escolas['MEDIO_nm'])

mean -0.000
std  0.132
min  -0.124
max  0.876


In [12]:
describe(escolas['TOT_l2'])

mean 0.012
std  0.007
min  0.000
max  0.043


#### RESPOSTA:<br> **LETRA B**

### Questão 4)
 
Utilize os atributos 'ENSINO MEDIO' e 'ANOS FINAIS'. Vamos transformá-los por meio da função  logaritmica. Para isso:

1. Faça uma cópia da base de dados, e atribua nulo (nan) a todos os valores iguais a zero nesses atributos,
2. Transforme esses atributos utilizando a operação da raiz quadrada e os adicione à base de dados com novos nomes, ex. sqrt(ENSINO MÉDIO) e  sqrt(ANOS FINAIS),
3. Remova todas as linhas que possuam nulo (nan) em qualquer um dos atributos transformados,
4. Utilizando $x$ como sqrt(ENSINO MÉDIO) e $y$ como sqrt(ANOS FINAIS), ajuste um modelo de regressão linear entre as duas variáveis. Use `LinearRegression()` do `sklearn`. Aqui não estamos interessados em separar treinamento e teste, só queremos ajustar uma função linear das duas variáveis, considerando todo os dados disponíveis.
5. Obtenha (imprima) o coeficiente da regressão linear aprendido, disponível no atributo coef_ do modelo inferido no passo anterior.

Qual o valor obtido do coeficiente no passo 5 acima, arredondado para 3 casas decimais?

(a) 0.481<br>
(b) 0.668<br>
(c) 0.690<br>
(d) 0.657<br>


In [13]:
escola_new = escolas.copy()

escola_new['ENSINO MEDIO'] = escola_new['ENSINO MEDIO'].replace(0, np.nan)
escola_new['ANOS FINAIS'] = escola_new['ANOS FINAIS'].replace(0, np.nan)

escola_new['ENSINO MEDIO SQRT'] = np.sqrt(escola_new['ENSINO MEDIO'])
escola_new['ANOS FINAIS SQRT'] = np.sqrt(escola_new['ANOS FINAIS'])

escola_new.dropna(subset=['ENSINO MEDIO SQRT', 'ANOS FINAIS SQRT'], inplace=True)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [14]:
X = escola_new['ENSINO MEDIO SQRT'].values.reshape(-1,1)
y = escola_new['ANOS FINAIS SQRT'].values.reshape(-1,1)

from sklearn.linear_model import LinearRegression

escola_ln = LinearRegression().fit(X, y)
coef = escola_ln.coef_
coef = coef[0][0]
print('COEFICICENTE: ', round(coef, 3))

COEFICICENTE:  0.668


#### RESPOSTA:<br> **LETRA B**

### Questão 5)
 
Utilize a base de dados antes da modificação feita na Questão 4. Codifique as variáveis categóricas 'MUN' (categórica nominal) e 'TOT_5I' (categórica ordinal).

Para MUN use números inteiros sequenciais, iniciados por 0 para codificar a variável em ordem alfabética, e gere um novo atributo MUN_cod

Para TOT_5I, use números inteiros sequenciais, iniciado por 0 para codificar a variável segundo sua ordenação, utilize para isso a ordenação crescente de seus valores únicos. Gere um novo atributo TOT_5I_ord.

A seguir, use a função `value_counts()` para mostrar a frequência de cada código na base de dados. Responda abaixo quais valores dos novos atributos (após codificação realizada) possuem a maior frequência (maior contagem):

(a) MUN_cod: código 1124;  TOT_5I: código 3095<br>
(b) MUN_cod: código 563;  TOT_5I: código 3<br>
(c) MUN_cod: código 1124;  TOT_5I: código 4<br>
(d) MUN_cod: código 563;  TOT_5I: código 4<br>

In [15]:
escolas['MUN_cod'] = escolas['MUN'].astype("category").cat.codes
escolas['TOT_5I_ord'] = escolas['TOT_5I'].astype("category").cat.codes


In [16]:
escolas['MUN_cod'].value_counts().idxmax()

563

In [17]:
escolas['TOT_5I_ord'].value_counts().idxmax()

4

#### RESPOSTA:<br> **LETRA D**